build: share Saucelabs browsers between karma test targets using background Saucelabs daemon and custom karma launcher

gregmagolan opened the pull request:
Picks up the work that @devversion started in the components repo.

Plan is to move this to the dev tools in a follow-up.

gregmagolan committed:
build: share Saucelabs browsers between karma test targets using background Saucelabs daemon and custom karma launcher

This upgrades the Saucelabs Bazel step on CI to use the more efficient Saucelabs daemon

gregmagolan commented on .circleci/config.yml:
```diff
@@ -675,6 +676,9 @@ workflows:
       - build-npm-packages:
           requires:
             - setup
+      - saucelabs:
```
Guessing enabling is for PRs will be in a follow-up?

devversion replied:
Yeah, let's do this in a follow-up.

gregmagolan replied:
✅

devversion commented on BUILD.bazel:
```diff
@@ -61,3 +70,11 @@ nodejs_binary(
+    name = "sauce_connect",
+    actual = select({
+        "@bazel_tools//src/conditions:linux": "@sauce_connect//:bin/sc",
+        "@bazel_tools//src/conditions:darwin": "@sauce_connect_mac//:bin/sc",
```
Nit: could use `@npm//@angular/build-tooling/constraints:linux_x64`, and `macos_x64`

gregmagolan replied:
✅

devversion commented on WORKSPACE:
```diff
@@ -213,3 +213,20 @@ register_toolchains(
+
+# Fetch sauce connect (tool to open Saucelabs tunnel for Saucelabs browser tests)
+http_archive(
+    name = "sauce_connect",
```
```suggestion
    name = "sauce_connect_linux_amd64",
```

gregmagolan replied:
✅

devversion commented on browser-providers.conf.d.ts:
```diff
@@ -0,0 +1,2 @@
+export const customLaunchers: any;
+export const sauceAliases: any;
```
Can this be typed properly?

gregmagolan replied:
✅

devversion commented on tools/saucelabs-daemon/README.md:
```diff
@@ -0,0 +1,78 @@
+
+3. In another terminal, you can run a particular test target through SauceLabs by prefixing the target name with "saucelabs_".
+
+For example, `packages/core/test:test_web` becomes `packages/core/test:saucelabs_test_web`.
```
I think we should move to a suffix. `test_web` and then `test_web_suffix`.

gregmagolan replied:
✅

devversion commented on tools/saucelabs-daemon/README.md:
```diff
@@ -0,0 +1,78 @@
+
+The karma_web_test rule for saucelabs must have a few important tags:
+
+*  `no-remote-exec` as they cannot be executed remotely since they require a local Sauce Connect process
```
```suggestion
*  `no-remote-exec` as they cannot be executed remotely since tests need to communicate with the daemon.
```

gregmagolan replied:
✅

devversion commented on tools/saucelabs-daemon/background-service/cli.ts:
```diff
@@ -0,0 +1,58 @@
+
+import {SaucelabsDaemon} from './saucelabs-daemon';
+
+const parallelExecutions = 2;
```
Nit: Should be an environment variable, with a default like in the circle comment.

gregmagolan replied:
I think it is better layers are a CLI argument. Changed to that:

```
yarn bazel run //tools/saucelabs-daemon/background-service -- ${NUMBER_OF_PARALLEL_BROWSERS}
```

WDYT?

devversion replied:
Yes

devversion commented on tools/saucelabs-daemon/background-service/cli.ts:
```diff
@@ -0,0 +1,58 @@
+}
+
+// First argument is the path to the sauce connect binary
+const sauceConnect = args[0];
```
Maybe validate that this is present & have a reasonable error then.

gregmagolan replied:
Good call. ✅

devversion commented on tools/saucelabs-daemon/background-service/ipc.ts:
```diff
@@ -0,0 +1,88 @@
+              .then((started) => {
+                if (!started) {
+                  console.debug('  > Browser not available.');
+                  this._noAvailableBrowser(socket);
```
```suggestion
                  this._sendUnavailableBrowserMessage(socket);
```

gregmagolan replied:
✅

devversion commented on tools/saucelabs-daemon/background-service/ipc.ts:
```diff
@@ -0,0 +1,88 @@
+                pageUrl: message.url,
+                requestedBrowserId: message.browserId,
+              })
+              .then((started) => {
```
Can we use `async/await` here?

gregmagolan replied:
Nice catch. Cleanup up.

devversion commented on tools/saucelabs-daemon/background-service/ipc.ts:
```diff
@@ -0,0 +1,88 @@
+          throw new Error(`Unsupported msg type: ${(message as any).type}`);
+      }
+    } catch (err) {
+      console.error((err as Error).message);
```
Let's just print the full error. 
```suggestion
      console.error(err);
```

gregmagolan replied:
💹

devversion commented on tools/saucelabs-daemon/background-service/ipc.ts:
```diff
@@ -0,0 +1,88 @@
+      }
+    } catch (err) {
+      console.error((err as Error).message);
+      this._internalError(socket, (err as Error).message);
```
And maybe here, conditionally have a proper `instanceof Error`, and if false, just use `err.toString()`

gregmagolan replied:
`err.toString()` sounds good to me since these are internal exceptions

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,306 @@
+        '--tunnel-identifier',
+        (this._userCapabilities as any).tunnelIdentifier || path.basename(tmpFolder),
+      ];
+      const sc = spawn(this._sauceConnect, sauceConnectArgs);
```
Can we extract the tunnel starting logic into a separate file & method?

devversion replied:
Also I think the tunnel should be started early on as it can take quite a while and we wouldn't want to shift all that time into individual bazel tests right?

gregmagolan replied:
Hmmm. In the happy path of all tests being cache hit and not test are run it would be wasteful starting the tunnel?

gregmagolan replied:
Split up into `_openSauceConnectTunnel` and `_launchBrowsers` anyway

devversion replied:
Yeah, I think it's a trade-off because now tests would need to wait for the tunnel to start, which could cause timeouts in the actual tests I'd assume. Starting the tunnel outside always is likely easier to follow/reason about

gregmagolan replied:
In all my testing the tunnel has opened up very quickly. Perhaps the problem is not as bad with the newest version of saucelabs connect or maybe Saucelabs has improved things on their end?

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,306 @@
+          // If a test has been scheduled before the browser completed launching, run
+          // it now given that the browser is ready now.
+          if (this._pendingTests.has(launched)) {
+            const _ = this._startBrowserTest(launched, this._pendingTests.get(launched)!);
```
```suggestion
this._startBrowserTest(launched, this._pendingTests.get(launched)!);
```

gregmagolan replied:
Ahh. Looks like tsetse doesn't like that which is why add added `const _ =`:

```
TS21225: [tsetse] All Promises in async functions must either be awaited or used in an expression
```

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,306 @@
+    if (!this._launching) {
+      // If this is the first test to run and we haven't launched any browsers yet
+      // then its now or never
+      await this.launchBrowsers().catch(err => {
```
So, let's say we just have a single karma test that needs to run. Right now we would launch X browsers, even though only 1x is needed.

gregmagolan replied:
Currently yes. The logic could be improved where the first test that runs starts the tunnel & launches one browser and then subsequent tests launch additional browsers up to the cap if there are none available.

gregmagolan replied:
Would be a bit of a refactor. Do this in a follow-up so its easier to follow and review that logic change?

devversion replied:
Sounds good to me.

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,306 @@
+        // Launch failed
+        throw this._launchError;
+      }
+    }
```
Should rather just use a promise that is stored and can be awaited here. It will reject with the error then as well?

gregmagolan replied:
Good call. Much cleaner that way.

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,306 @@
+    this._runningTests.delete(testId);
+  }
+
+  async startTest(test: BrowserTest): Promise<boolean> {
```
Nit: can we add a description to these functions. Maybe mentioning like what the boolean is meant to do.

gregmagolan replied:
✅ Added descriptions to the internal functions as well.

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,306 @@
+        }
+      }
+
+      const _ = this._startBrowserTest(browser, test);
```
```suggestion
      this._startBrowserTest(browser, test);
```

gregmagolan replied:
✅

devversion commented on tools/saucelabs-daemon/launcher/BUILD.bazel:
```diff
@@ -0,0 +1,24 @@
+    ],
+)
+
+esbuild(
```
Add a comment on why this is built as CommonJS.

gregmagolan replied:
✅

devversion commented on tools/saucelabs-daemon/launcher/launcher.ts:
```diff
@@ -0,0 +1,83 @@
+
+  this.on('kill', async (doneFn: () => void) => {
+    _endBrowserTest();
+    daemonConnection!.end();
```
```suggestion
    daemonConnection?.end();
```

gregmagolan replied:
✅

devversion reviewed (approved):
Overall LGTM- but I'm definitely biased a bit 😄 

Next steps would be:

1. Addressing the feedback & landing this
2. Making sure all parts are tested, like with the legacy job
3. Replace the legacy job & delete code, in favor of just the Bazel job

JiaLiPassion reviewed (approved):
LGTM!
Reviewed-for: fw-zones

gregmagolan commented on tools/saucelabs-daemon/README.md:
```diff
@@ -0,0 +1,78 @@
+To see the test output while the tests are running (as these are long tests), add the `--test_output=streamed` option. 
+Note, this option will also prevent bazel from using the test cache and will force the test to run.
+
+`bazel query` is required gather up all karma saucelabs test labels so they can be run in one command as they are tagged `manual`.
```
✅

gregmagolan commented on tools/saucelabs-daemon/background-service/cli.ts:
```diff
@@ -0,0 +1,58 @@
+const accessKey = process.env.SAUCE_ACCESS_KEY;
+const tunnelIdentifier = process.env.SAUCE_TUNNEL_IDENTIFIER;
+
+const buildName = process.env.CIRCLECI ? process.env.CIRCLE_BUILD_NUM : 'localdev';
```
✅

gregmagolan commented on modules/playground/e2e_test/async/async_spec.ts:
```diff
@@ -69,8 +69,9 @@ describe('async', () => {
-      return browser.executeAsyncScript('window.frameworkStabilizers[0](arguments[0]);');
+    const whenAllStable = (): Promise<any> => {
+      return browser.executeAsyncScript('window.frameworkStabilizers[0](arguments[0]);') as
+          Promise<any>;
```
Nice catch. I think that snuck in when I was trying to upgrade to selenium v4. Removed the change.

gregmagolan commented on tools/saucelabs-daemon/README.md:
````diff
@@ -0,0 +1,78 @@
+
+``` bash
+yarn bazel run //tools/saucelabs:sauce_service_setup
+TESTS=$(./node_modules/.bin/bazelisk query --output label '(kind(karma_web_test, ...) intersect attr("tags", "saucelabs", ...)) except attr("tags", "fixme-saucelabs", ...)')
````
New `/scripts/saucelabs` ?

gregmagolan replied:
✅

gregmagolan commented on tools/saucelabs-daemon/README.md:
````diff
@@ -0,0 +1,78 @@
+``` bash
+yarn bazel run //tools/saucelabs:sauce_service_setup
+TESTS=$(./node_modules/.bin/bazelisk query --output label '(kind(karma_web_test, ...) intersect attr("tags", "saucelabs", ...)) except attr("tags", "fixme-saucelabs", ...)')
+yarn bazel test --config=saucelabs ${TESTS}
````
Nice catch. That is key. I'll update it to use the environment variable as well.

gregmagolan replied:
Actually, I opted for a CLI parameter since users will generally launch the background service in one terminal and Bazel in another.

gregmagolan commented on .circleci/config.yml:
```diff
@@ -313,14 +313,15 @@ jobs:
             TESTS=$(./node_modules/.bin/bazelisk query --output label '(kind(karma_web_test, ...) intersect attr("tags", "saucelabs", ...)) except attr("tags", "fixme-saucelabs", ...)')
-            yarn bazel test --config=saucelabs ${TESTS}
-            yarn bazel run //tools/saucelabs:sauce_service_stop
+            yarn bazel test --config=saucelabs --jobs=2 ${TESTS}
```
How's this
```
    environment:
      NUMBER_OF_PARALLEL_BROWSERS: 2
```

dgp1130 reviewed (approved):
Reviewed-for: code-ownership

I'll defer to others on the actual implementation, but I can at least LGTM the PullApprove change.

devversion commented on tools/saucelabs-daemon/background-service/cli.ts:
```diff
@@ -0,0 +1,70 @@
+  throw Error('No tunnel set up. Please set the `SAUCE_TUNNEL_IDENTIFIER` variable.');
+}
+
+if (!buildName) {
```
This can never be false I think

gregmagolan replied:
nice catch

devversion commented on tools/saucelabs-daemon/background-service/ipc.ts:
```diff
@@ -0,0 +1,79 @@
+
+let nextSocketId = 0;
+
+export class IpcServer {
```
Back from prototyping we didn't have any comments here, but should we have super small ones here for best practices? just high-level explaining what we intend to do

gregmagolan replied:
Added some comments to this class and the SaucelabsDaemon class

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,356 @@
+
+interface RemoteBrowser {
+  id: string;
+  state: 'acquired'|'free'|'launching';
```
I know it's my fault, but can we maybe rename to `claimed` instead of `acquired`? I will let you decide as native english speaker here though.

gregmagolan replied:
"claimed" sounds good to me. I can see how `acquired` is ambiguous in the saucelabs context. "in_use" or "reserved" are two other alternatives but will go with "claimed"

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,356 @@
+   * This is typically done when the first test is started so that no connection is made
+   * if all tests are cache hits.
+   */
+  async connect() {
```
```suggestion
  async connectTunnel() {
```

gregmagolan replied:
👍

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,356 @@
+  /** Id of the keep alive interval that ensures no remote browsers time out. */
+  private _keepAliveIntervalId: NodeJS.Timeout|null = null;
+
+  /* Have we connected to Saucelabs or are in the process of connecting? */
+  private _connection: Promise<void>|undefined = undefined;
```
```suggestion
  /* Promise  indicating whether we the tunnel is active, or if we are still connecting. */
  private _tunnelConnection: Promise<void>|undefined = undefined;
```

gregmagolan replied:
👍

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,356 @@
+   * this initiates that process and awaits until it succeeds or fails.
+   *
+   * If the daemon has already initiated the saucelabs tunnel creation and browser launching then
+   * but it is not yet complete then this until it succeeds or fails.
```
and browser launching is not yet complete? this phrase is incomplete/not accurate I feel like

gregmagolan replied:
fixed up

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,356 @@
+   * @internal
+   * Establishes the Saucelabs connect tunnel.
+   **/
+  async _openSauceConnectTunnel() {
```
Why can't this be a function in a separate file to not bloat this daemon?

gregmagolan replied:
How would you split it up? These are private functions of the SaucelabsDaemonn class

gregmagolan replied:
✅

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,356 @@
+
+      // TS21225: [tsetse] All Promises in async functions must either be awaited or used in an
+      // expression
+      const _ = this._startBrowserTest(browser, test);
```
Why aren't we awaiting this here? We should note that in the comment as well, or change

gregmagolan replied:
Commented here. This function really only cares that the browser test start was initiated not that it was started successfully. Failures should be caught by the karma test itself either as a failure or a timeout.

devversion commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -0,0 +1,356 @@
+          // it now given that the browser is ready now.
+          if (this._pendingTests.has(launched)) {
+            // TS21225: [tsetse] All Promises in async functions must either be awaited or used in
+            // an expression
```
Nit: Add a note why we are not interested in awaiting here

devversion replied:
Also need to be careful of errors here with loading the test browser page right?

gregmagolan replied:
In the daemon we'll just ignore those errors.. tho it will log
```
    } catch (e) {
      console.error('Could not start browser test with id', test.testId, test.pageUrl);
    }
```

The safety is that the karma test should timeout or fail if that happens.

devversion reviewed (approved):
LGTM. Couple of more comments, but close. Can you please push feedback as fixup commits?

gregmagolan committed:
build: address review feedback; added scripts/test/run-saucelabs-tests.sh script for local testing

more chars to meet the linters requirements

gregmagolan commented:
> Overall LGTM- but I'm definitely biased a bit 😄
> 
> Next steps would be:
> 
> 1. Addressing the feedback & landing this
> 2. Making sure all parts are tested, like with the legacy job
> 3. Replace the legacy job & delete code, in favor of just the Bazel job

Ready for another pass @devversion. Changes since last review: https://github.com/angular/angular/pull/49200/commits/e687ce2ec5dc5a766e494222ca5f1525b85f257e

devversion reviewed (approved):
LGTM. final comments. thanks Greg!

gregmagolan commented on tools/saucelabs-daemon/background-service/saucelabs-daemon.ts:
```diff
@@ -307,20 +317,26 @@ export class SaucelabsDaemon {
+    browser.state = 'claimed';
+
+    console.debug(`Opening test url for #${test.testId}: ${test.pageUrl}`);
+    browser.driver!.get(test.pageUrl)
```
nice trick 👍

gregmagolan committed:
build: address review feedback; should be ready to land now... additional chars to meet commit msg formatting requirements

plus more additional chars here

gregmagolan commented on scripts/test/run-saucelabs-tests.sh:
```diff
@@ -0,0 +1,51 @@
+function kill_background_service {
+  echo "Killing background service..."
+  kill $BACKGROUND_SERVICE_PID  # Kill the backgound service
+  sleep 2  # Let the output of the background service flush
```
👍

gregmagolan commented on scripts/test/run-saucelabs-tests.sh:
```diff
@@ -0,0 +1,51 @@
+trap kill_background_service INT TERM
+
+# Small pause to give time for the background service to open up its IPC port and start listening
+sleep 2
```
Yeah. Didn't think it was worth extra complexity to waiting the port.

gregmagolan committed:
build: don't run saucelabs tests yet on PRs... that will happen in a followup

additional test to make linter happy

gregmagolan commented:
Last CI run with saucelabs tests enabled on PRs with passing tests: https://app.circleci.com/pipelines/github/angular/angular/60980/workflows/049e20fc-1957-45c2-b21c-4b0908577ee5/jobs/1335782

Removing the job from this PR now for landing as enabling it will be in a follow-up

devversion commented on scripts/test/run-saucelabs-tests.sh:
```diff
@@ -21,23 +21,27 @@ if [[ -z "${SAUCE_TUNNEL_IDENTIFIER:-}" ]]; then
+# Start the saucelabs-daemon background service in the background. Run directly from the generated
+# bash script instead of using bazel run so we get the PID of the node process. Otherwise killing
+# the child process in kill_background_service doesn't kill the spawn node process.
+cd dist/bin/tools/saucelabs-daemon/background-service/background-service.sh.runfiles/angular
```
I'm good with the code as is, but it seems a little unexpected to me. the Node process should be a child of the `bazel run` process, so it should be killed.

gregmagolan replied:
It was surprising to me as well. Could use some more investigation in the future.

devversion replied:
this is just the convenience script anyway right?

gregmagolan replied:
yes, this is not used on CI at all

devversion reviewed (approved):
LGTM

Reviewed-for: global-dev-infra-approvers

AndrewKushnir commented:
This PR was merged into the repository by commit d0262cd0e4463fba9ccf8b421fcf1a7ddd668582.
